common/chaton.hpp

```diff
@@ -0,0 +1,758 @@
+ *   * [tagged msg] the string containing the tagged message(s)
+ *   * [parts lengths] an array of integers, which specifies the part lengths,
+ *     which divides the returned string into parts.
+ *   * [parts types] a string where each character indicates whether the corresponding
+ *     part is a normal part which needs to be tokenized without parse_special
+ *     or is a special part which needs to be tokenized with parse-special.
+ * 
+ * 
+ * ## example/main
+ * 
+ * The interactive commandline program under example/main, uses
+ * 
+ * * the system role related tags to tag the system prompt
+ *   * the system prompt includes contents of -p if any
+ *   * followed by contents of file specified using -f if any
+ * * the user begin+prefix to map to in-prefix
+ * * the user suffix+end to map to in-suffix
+ * * the reverse-prompt to map to antiprompt
+ * * wrt tokenization
+ *   * the user specified system prompt is tokenized with parse_special flag.
+ *   * however the user messages are tokenized without parse_special flag.
+ * 
+ * Currently Main doesnt use chaton-tmpl-apply, but only 
+ * * chaton-tmpl-apply-single (for system prompt) and
+ * * chaton-tmpl-role-kv which maps the user prefix, suffix and reverse-prompt
+ *   to in-prefix, in-suffix and antiprompt of main.
+ * These always adds any role specific begin+prefix and suffix+end around
+ * the passed message.
+ * 
+ * ## other uses be it wrt llama.cpp-as-library or examples/server or ...
+ * 
+ * This module exposes a c-api which is equivalent to the current hardcoded
+ * templating logic's llama_chat_apply_template. So any program using llama.cpp's
+ * chat templating logic can be easily migrated to make use of this generic code
+ * with text based config file based flow.
+ * 
+ * If a program doesnt want to bring in json dependency into their project,
+ * there is also common/simpcfg.hpp, which provides a simple text based config
+ * file format, along with the corresponding parser for the same. This can be
+ * modified to work with simpcfg easily, if needed.
+ * 
+ * ## Adding support for new model / chat-handshake-template-standard
+ * 
+ * 1. Add suitable entries in json for that model/standard
+ *    This in itself should work for most of the models.
+ * 
+ * 2. If some new model introduces a totally different kind of chat-templating
+ *    tag inter/intra mixing, Try to reuse and update the generic flow in
+ *    chaton-tmpl-apply, as much as possible, before trying to add any custom logic.
+ * 
+ *    If you update the generic flow, cross check if existing json files will
+ *    need to be updated or not.
+ * 
+ * 
+ * ## Notes
+ * 
+ * Look at the sample chaton_meta.json in examples folder for how the above may apply to
+ * the different llm's out there like
+ * 
+ * * llama2, llama3, gemma, zephyr, deepseek(normal and coder), monarch, mistral, phi3
+ * * chatml, command-r, orion, openchat, vicuna
+ * 
+ */
+
+#include <string>
+#include <fstream>
+#include <iostream>
```

Comment:
We generally avoid `<iostream>` - use the available `LOG_` functionality in the examples

Reply:
iostream is a vestiage from my initial experimentation wrt this flow, currently wrt chaton.hpp, I dont use it and I forgot to remove it.

However wrt simpcfg.hpp, I use it has part of its standalone testing, however for the integrated into llama.cpp, it is avoidable.